File: pipeline/models/alignments.py
"""Pydantic model for alignments.json."""

from __future__ import annotations

from .common import LenientModel, StrictModel


class PantheonInfo(StrictModel):
    """A pantheon grouping."""

    name: str
    description: str


class AlignmentSin(StrictModel):
    """A sin entry with devotion threshold."""

    devotion: int
    sin: str


class Alignment(StrictModel):
    """A single alignment/deity."""

    id: str
    name: str
    pantheon: str
    concepts: list[str]
    description: str
    commandments: list[str]
    restriction: str | None = None
    sins: list[AlignmentSin]


class AlignmentsFile(LenientModel):
    """Top-level structure of alignments.json."""

    pantheons: dict[str, PantheonInfo]
    alignments: list[Alignment]
    devotionMechanics: dict | None = None

File: src/lib/dtd/schemas/alignments.ts
/**
 * Zod schema for alignments data — replaces former Pydantic model.
 */

import { z } from "zod";

// ---------------------------------------------------------------------------
// PantheonInfo (strict)
// ---------------------------------------------------------------------------

export const PantheonInfo = z
	.object({
		name: z.string(),
		description: z.string(),
	})
	.strict();
export type PantheonInfo = z.infer<typeof PantheonInfo>;

// ---------------------------------------------------------------------------
// AlignmentSin (strict)
// ---------------------------------------------------------------------------

export const AlignmentSin = z
	.object({
		devotion: z.number().int(),
		sin: z.string(),
	})
	.strict();
export type AlignmentSin = z.infer<typeof AlignmentSin>;

// ---------------------------------------------------------------------------
// Alignment (strict)
// ---------------------------------------------------------------------------

export const Alignment = z
	.object({
		id: z.string(),
		name: z.string(),
		pantheon: z.string(),
		concepts: z.array(z.string()),
		description: z.string(),
		commandments: z.array(z.string()),
		restriction: z.string().nullish(),
		sins: z.array(AlignmentSin),
	})
	.strict();
export type Alignment = z.infer<typeof Alignment>;

// ---------------------------------------------------------------------------
// AlignmentsFile (lenient)
// ---------------------------------------------------------------------------

export const AlignmentsFile = z
	.object({
		pantheons: z.record(z.string(), PantheonInfo),
		alignments: z.array(Alignment),
		devotionMechanics: z.record(z.string(), z.unknown()).nullish(),
	})
	.passthrough();
export type AlignmentsFile = z.infer<typeof AlignmentsFile>;
